File: apps/mobile/src/features/core/native-core-adapters/recommendation.ts
import type { FeedbackEvent, Recommendation } from '@glimpse/shared';

import { nativeCoreBridgeHelpers } from '../native-core-bridge';
import type { BridgeCoreClient } from '../types';

import type { NativeAdapterDeps } from './common';

const {
  fromNitroRecommendation,
  toNitroFeedbackEvent,
  toNitroRecommendation,
} = nativeCoreBridgeHelpers;

export function createRecommendationAdapter(
  deps: NativeAdapterDeps,
): Pick<
  BridgeCoreClient,
  | 'saveRecommendations'
  | 'listRecommendations'
  | 'listPendingRecommendations'
  | 'respondToRecommendation'
> {
  const { fallbackClient, runCoreAsync } = deps;

  return {
    async saveRecommendations(recommendations: Recommendation[]): Promise<void> {
      await runCoreAsync(
        (client) => client.saveRecommendations(recommendations.map(toNitroRecommendation)),
        () => fallbackClient.saveRecommendations(recommendations),
      );
    },

    async listRecommendations(): Promise<Recommendation[]> {
      return runCoreAsync(
        async (client) => (await client.listRecommendations()).map(fromNitroRecommendation),
        () => fallbackClient.listRecommendations(),
      );
    },

    async listPendingRecommendations(): Promise<Recommendation[]> {
      return runCoreAsync(
        async (client) =>
          (await client.listPendingRecommendations()).map(fromNitroRecommendation),
        () => fallbackClient.listPendingRecommendations(),
      );
    },

    async respondToRecommendation(
      recommendationId: string,
      status: 'accepted' | 'ignored' | 'dismissed',
      feedbackEvent: FeedbackEvent,
    ): Promise<void> {
      await runCoreAsync<void>(
        (client) =>
          client.respondToRecommendation(
            recommendationId,
            status,
            toNitroFeedbackEvent(feedbackEvent),
          ),
        () => fallbackClient.respondToRecommendation(recommendationId, status, feedbackEvent),
      );
    },
  };
}

File: packages/core-rust/src/application/recommendation.rs
use crate::error::Result;
use crate::models::{
    FeedbackEvent, GraphAnalysisCommitResult, GraphAnalysisRecord, Recommendation,
    RecommendationStatus,
};

use super::SharedCore;

impl SharedCore {
    pub fn list_graph_analysis_records(&self) -> Result<Vec<GraphAnalysisRecord>> {
        self.client().list_graph_analysis_records()
    }

    pub fn commit_graph_analysis(
        &self,
        records: &[GraphAnalysisRecord],
        recommendations: &[Recommendation],
    ) -> Result<GraphAnalysisCommitResult> {
        self.client()
            .commit_graph_analysis(records, recommendations)
    }

    pub fn save_recommendations(&self, recommendations: &[Recommendation]) -> Result<()> {
        self.client().save_recommendations(recommendations)
    }

    pub fn list_recommendations(&self) -> Result<Vec<Recommendation>> {
        self.client().list_recommendations()
    }

    pub fn list_pending_recommendations(&self) -> Result<Vec<Recommendation>> {
        self.client().list_pending_recommendations()
    }

    pub fn respond_to_recommendation(
        &self,
        recommendation_id: &str,
        status: RecommendationStatus,
        feedback_event: &FeedbackEvent,
    ) -> Result<()> {
        self.client()
            .respond_to_recommendation(recommendation_id, status, feedback_event)
    }
}
